### PR TITLE
Add custom grid helper based on CSS custom properties

### PR DESCRIPTION
**Pull request by lucasvallenet**
Grid Helper is often used in our projects to make sure the layout is aligned with the design. Here is a fully customizable version with no CSS dependencies appart from the base CSS Variables located at `/assets/styles/elements/_page.scss`.
It allows us to customize grid gutters, columns and margins with CSS only.

**Use:**
After the `<body>` tag write the following line:
```
<div data-module-grid-helper></div>
```

You can hardcode values directly in the html:
```
<div 
    data-module-grid-helper
    style="--grid-columns: 8; --grid-gutter: 20px;"
></div>
```

Or use the CSS in `/assets/styles/elements/_page.scss`:
```
:root {

    // Grid
    --grid-columns      : 4; // Number of columns
    --grid-gutter       : #{rem(10px)}; // Gutter size
    --grid-gutter-half  : calc(0.5 * var(--grid-gutter)); // Half gutter size often use in CSS
    --grid-margin       : 0px; // Grid horizontal margin

    // Update any of these vars for a responsive grid-helper
    @media (min-width: $from-small) {
        --grid-columns  : 12;
        --grid-gutter   : #{rem(16px)};
        --grid-margin   : #{rem(20px)};
    }
}
```

Press `CTRL + G` to toggle the grid helper.

### CODE REVIEW

**Comment by devenini** [comment]
I like the idea. However I'm not a fan of having this in the boilerplate without having an environment variables setup to make sure that these types of modules are not included on a live production. Like a "debug/development" global variable.

**Comment by lucasvallenet** [comment]
We could use the `env.process.NODE_ENV` that return `development` on watch and `production` on build, but I can't get around the ES6 synthax to conditionally import/export the component. Any idea @mcaskill ?

**Comment by mcaskill** [comment]
Hrm. We would need to update the `build/tasks/styles.js` and implement a feature similar to the Webpack sass-loader's [`additionaldata`](https://github.com/webpack-contrib/sass-loader/tree/v12.6.0#additionaldata) option. Not sure how that feature works though.

**Comment by lucasvallenet** [comment]
@mcaskill the main issue here is not for css but for javascript.

We'd like to import the component only on development, unfortunatly this is not doable:
```
if(process.env.NODE_ENV === 'development') {
    export { default as GridHelper } from './modules/GridHelper';
}
```

Could we bundle a development file for js utilities we use for development only, like vendors ? And then find a way to only include it on development environment in the frontend like so:
```
<script src="{{# withBaseUrl }}assets/scripts/develoment.js?v={{ assetsVersion }}{{/ withBaseUrl }}"></script>
```

**Comment by mcaskill** [comment]
Thanks, I better understand the situation. There must be a way with BrowserSync to inject additional `<script>` elements into a page, like it does (I presume) for its live-reloading.
I'd prefer if our development tools be a separate JS entry file to avoid polluting the project's primary entry file.

File: assets/scripts/modules/GridHelper.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        elStyles.zIndex = '10000'
+        elStyles.position = 'fixed'
+        elStyles.top = '0'
+        elStyles.left = '0'
+        elStyles.display = 'flex'
+        elStyles.width = '100%'
+        elStyles.height = '100%'
+        elStyles.columnGap = GridHelper.settings.GUTTER
+        elStyles.paddingLeft = `calc(${GridHelper.settings.MARGIN} + ${GridHelper.settings.GUTTER})`
+        elStyles.paddingRight = `calc(${GridHelper.settings.MARGIN} + ${GridHelper.settings.GUTTER})`
```
Comment (Jerek0) [inline assets/scripts/modules/GridHelper.js:53]: I think it should be only `GridHelper.settings.MARGIN` for lateral paddings?

Comment (devenini) [reply assets/scripts/modules/GridHelper.js:53]: Should the GridHelper use grid css instead of flex and padding? It would be coherent with #108

Comment (lucasvallenet) [reply assets/scripts/modules/GridHelper.js:53]: The main issue with the grid display is that it doesn't fit with the actual design grid system if we use `fr` units, that's why I used flexbox

**Comment by Jerek0** [comment]
We could remove the import in `modules.js` and the grid helper markup by default and use it cautiously only when we need for now? 
That would allow us to merge this PR and we could tackle the development-only scripts concept in a new PR?

**Review by Jerek0 — APPROVED** [review]

**Comment by mcaskill** [comment]
#### Outcome from meeting on 2022-10-05

* Refactor to be a standalone utility to be included only in dev mode.
* Move markup currently in HTML documents into the utility.
* Figure out how to bundle the utility automatically during development builds (via ESBuild).
* Figure out how to inject "is debug" flag into the `window` in development builds (via ESBuild).

**Comment by arnvvd** [comment]
**Updates**
- I rewrote the grid helper as a standalone utility.
- I used a dynamic import for the grid helpers
```
// Dynamic imports for development mode only
let gridHelper;
(async () => {
    if (config.IS_DEV) {
        const gridHelperModule = await import('./utils/grid-helper');
        gridHelper = gridHelperModule?.gridHelper;
    }
})();

function init() {
    /**
     * Add grid helper
     */
    gridHelper?.();
}
```

- You can now use the `$app-env` sass variable to use conditional styles in your .scss files
```
@if $app-env == development {
    .myDiv {
         color: red;
    }
}
  ```

**Review by mcaskill — APPROVED** [review]
Looks good to me but I have not tested this.

For future reference, when working on a feature branch and pulling in latest changes from the production branch, use `git rebase` instead of `git merge`. This will make for a cleaner history during the development of a feature branch and once its merged into the production branch.

**Review by devenini — APPROVED** [review]

**Review by plcossette — APPROVED** [review]

**Review by Jerek0 — APPROVED** [review]

**Pull request merged**